### PR TITLE
G2 2024 v5 issue#15271 Create a retrieveShowDuggaService microservice

### PR DESCRIPTION
**Pull request by a22albjo**
To test 
Add this to the includes in a microservice for example saveDugga_ms.php:
`include_once "retrieveShowDuggaService.php";`

Then add this at the end of the file:
```
retrieveShowDuggaService(
	$debug,
	$param,
	$answer,
	$variantanswer,
	$score,
	$highscoremode,
	$grade,
	$submitted,
	$marked,
	$duggainfo,
	$files,
	$userfeedback,
	$feedbackquestion,
	$savedvariant,
	$ishashindb,
	$variantsize,
	$variantvalue,
	$password,
	$hashvariant,
	$isFileSubmitted,
	$isTeacher, // isTeacher is true for both teachers and superusers
	$variants,
	$duggatitle,
	$hash,
	$hashpwd,
	$opt,
	$link,
	$active,
	$group);
```

Update dugga.js to use the ms instead of the mono, like this for save dugga service:
```
else if(kind=="PDUGGA"){
		if(strcmp($opt,"SAVDU")==0){
			$.ajax({
				url: "../DuggaSys/microservices/showDuggaService/saveDugga_ms.php",
				type: "POST",
				data: "courseid="+querystring['cid']+"&did="+querystring['did']+"&coursevers="+querystring['coursevers']+"&moment="+querystring['moment']+"&segment="+querystring['segment']+"&hash="+hash+"&password="+pwd+"&opt="+opt+para+"&variant="+variantValue,
				datatype: "json",
				success: returnedDugga
		})}
		$.ajax({
			url: "showDuggaservice.php",
			type: "POST",
			data: "courseid="+querystring['cid']+"&did="+querystring['did']+"&coursevers="+querystring['coursevers']+"&moment="+querystring['moment']+"&segment="+querystring['segment']+"&hash="+hash+"&password="+pwd+"&opt="+opt+para+"&variant="+variantValue,
			datatype: "json",
			success: returnedDugga
		});
	}
```

Then try to save a dugga and enjoy the output.

### CODE REVIEW

**Review by a22marca — CHANGES_REQUESTED** [review]
We might want to change it so the retrieveShowDuggaService to return the array instead of the encoded json.

There is a check using the functions checklogin and isSuperUser in showDuggaService under the comment `// Retrieve information` in showDuggaservice.php, that we might want to include in this microservice?

I used this code when testing the integration:
```
	}else if(kind=="PDUGGA"){
		let service_url = "showDuggaservice.php";
		if(opt=="SAVDU"){
			service_url = "../DuggaSys/microservices/showDuggaService/saveDugga_ms.php";
		}
		$.ajax({
			url: service_url,
			type: "POST",
			data: "courseid="+querystring['cid']+"&did="+querystring['did']+"&coursevers="+querystring['coursevers']+"&moment="+querystring['moment']+"&segment="+querystring['segment']+"&hash="+hash+"&password="+pwd+"&opt="+opt+para+"&variant="+variantValue,
			datatype: "json",
			success: returnedDugga
		});
```

For me the response seems correct. But there seems to be a few issues with the microservice I used for testing (saveDugga). It does not save the actual dugga, no hash or password is being sent when saving. Maybe it's on my end, but maybe you could check again? This is the debug response:
```
debug: "Unable to save dugga!"
```

**Comment by a22albjo** [comment]
You can use this code to test:
```
else if(kind=="PDUGGA"){
		let service_url = "showDuggaservice.php";
		if(opt=="SAVDU"){
			service_url = "../DuggaSys/microservices/showDuggaService/saveDugga_ms.php";
		}
		if(opt=="UPDATEAU"){
			service_url = "../DuggaSys/microservices/showDuggaService/updateActiveUsers_ms.php";
		}
		$.ajax({
			url: service_url,
			type: "POST",
			data: "courseid="+querystring['cid']+"&did="+querystring['did']+"&coursevers="+querystring['coursevers']+"&moment="+querystring['moment']+"&segment="+querystring['segment']+"&hash="+hash+"&password="+pwd+"&opt="+opt+para+"&variant="+variantValue,
			datatype: "json",
			success: returnedDugga
		});
	}
```

File: DuggaSys/microservices/showDuggaService/updateActiveUsers_ms.php [hunk 1]
```diff
@@ -27,4 +29,33 @@
     $query->execute();
 }
 
-echo json_encode($active);
+echo json_encode(retrieveShowDuggaService(
```
Comment (a22marca) [inline DuggaSys/microservices/showDuggaService/updateActiveUsers_ms.php:32]: Declare the variables that are used as parameters here

File: Shared/dugga.js [hunk 2]
```diff
@@ -1100,29 +1100,12 @@ function AJAXService(opt,apara,kind)
 				success: returnedSection
 			});
 	}else if(kind=="PDUGGA"){
-		//for testing
-		console.log("dugga.js (kind=='PDUGGA')\n"+"courseid="+querystring['courseid']+"\ndid="+querystring['did']+"\ncoursevers="+querystring['coursevers']+"\nmoment="+querystring['moment']+"\nsegment="+querystring['segment']);
```
Comment (a22marca) [inline Shared/dugga.js:1104]: We should add a switch for the opts here

**Review by a22marca — CHANGES_REQUESTED** [review]
- Savedugga seems to work well
- I'm not really sure where loaddugga and updateActiveUsers is used on the website itself (if it can be used at all?)

See my other comments for needed changes

File: Shared/dugga.js (hunk 2 above)
Comment (a22albjo) [reply Shared/dugga.js:1104]: Commit 75eb68e contains the switch statements, i have also created a new issue "migrate showDuggaSevice.php" where implementing the switch statement is a part of the issue.

File: DuggaSys/microservices/showDuggaService/updateActiveUsers_ms.php (hunk 1 above)
Comment (a22albjo) [reply DuggaSys/microservices/showDuggaService/updateActiveUsers_ms.php:89]: Added the variable assignments from the monorepo to ensure full coverage of all the variables.

**Comment by a22albjo** [comment]
I have also created a getShowDugga_ms.php. It is an important part of the retriveimplementation to make it possible to run the retriveShowDuggaService_ms.php without running any service before it.

**Review by a22marca — APPROVED** [review]
Everything seems to be addressed, and the switch case will be added in a separate issue.

- loaddugga is performed by retrieve show dugga according to @a22albjo 
- updateActiveUsers does not seem to be used on the website

I think this is ok to merge now

**Pull request merged**